reasoner/Reasoner.java

```diff
@@ -77,56 +83,102 @@ ResolverRegistry resolverRegistry() {
-        if (disjunction.conjunctions().size() > 1) answers = answers.distinct();
+        if (satisfiableDisjunction.conjunctions().size() > 1) answers = answers.distinct();
         return answers;
     }
 
-    private Producer<ConceptMap> producer(Conjunction conjunction) {
-        return producer(conjunction, set(), defaultContext);
+    private boolean isInfer(Disjunction disjunction, Context.Query context) {
+        if (!context.options().infer() || context.transactionType().isWrite() || !logicMgr.rules().hasNext()) {
+            return false;
+        }
+        for (Conjunction conj : disjunction.conjunctions()) {
+            for (Variable var : conj.variables()) {
+                if (var.resolvedTypes().isEmpty()) return true;
+                for (Label type : var.resolvedTypes()) {
+                    if (logicMgr.rulesConcluding(type).hasNext() || logicMgr.rulesConcludingHas(type).hasNext()) {
+                        return true;
+                    }
+                }
+            }
+            if (!conj.negations().isEmpty()) {
+                for (Negation negation : conj.negations()) {
+                    if (isInfer(negation.disjunction(), context)){
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
     }
 
-    private Producer<ConceptMap> producer(Conjunction conjunction, Set<Identifier.Variable.Name> filter,
-                                          Context.Query context) {
-        Producer<ConceptMap> producer;
-        logicMgr.typeResolver().resolve(conjunction);
-        if (conjunction.isSatisfiable()) {
-            if (isInfer(context)) producer = resolve(conjunction, filter);
-            else producer = traversalEng.producer(
-                    conjunction.traversal(filter), context.producer(), PARALLELISATION_FACTOR
-            ).map(conceptMgr::conceptMap);
-        } else if (!conjunction.isBounded() && conjunctionContainsThings(conjunction, filter)) {
-            throw GraknException.of(UNSATISFIABLE_CONJUNCTION, conjunction);
+    private ResourceIterator<ConceptMap> resolve(Disjunction disjunction, Set<Identifier.Variable.Name> filter, Long offset, Long limit, Context.Query context) {
+        Set<Reference.Name> nameFilter = iterate(filter).map(Identifier.Variable.Name::reference).toSet();
+        if (disjunction.conjunctions().size() == 1) {
+            return produce(new ReasonerProducer(disjunction.conjunctions().get(0), resolverRegistry, nameFilter, offset, limit), context.producer(), asyncPool1());
         } else {
-            return Producers.empty();
+            return produce(new ReasonerProducer(disjunction, resolverRegistry, nameFilter, offset, limit), context.producer(), asyncPool1());
+        }
+    }
+
+    private Disjunction resolveTypesAndFilter(Disjunction disjunction, Set<Identifier.Variable.Name> filter) {
```

Comment:
Throw an exception if a non-bounded conjunction is not satisfiable, and never throw an exception for a schema query (that's why we check `conjunctionContainsThing()`.